[Chung Loong] iP

Hikoya opened the pull request:
### Skylark

> "You can do anything, but not everything." – David Allen

Skylark frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/Hikoya/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE!**

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the main method:

```
public class Skylark {
    public static void main(String[] args) {
        Skylark skylark = new Skylark();
        skylark.startRunning();
    }
}

```

Commit:
Add Gradle support

Hikoya committed:
Rename Duke to Skylark

Making my chatbot unique by giving it a name first

Hikoya committed:
Complete Level-1 Greet, Echo, Exit

Hikoya committed:
Complete Level 2. Add, List

Hikoya committed:
Add a new Task.java class

This Task class is used to track status of a particular Task

Hikoya committed:
Update Task.java and add Done and Undone

Hikoya committed:
Complete Level 3. Mark as Done

Hikoya committed:
Create toString method for Task

Hikoya committed:
Create Event, ToDo and Deadline class

Hikoya committed:
Complete Level 4. ToDos, Events, Deadlines

Hikoya committed:
Adding in input test case for automated testing

Hikoya committed:
Create SkylarkException and ToDoException

Hikoya committed:
Move main logic into separate function that throws Exception

Hikoya committed:
Complete Level 5. Handle Errors

Teach Skylark to deal with errors such as incorrect inputs entered by the user.

Hikoya committed:
Change subString function to startsWith

Hikoya committed:
Complete Level 6. Delete

Add support for deleting tasks from the list.

Hikoya committed:
Moved commands into a single Enum

Hikoya committed:
Added new Exceptions to check for edge cases

Hikoya committed:
Updated input.txt for level-6

Hikoya committed:
Make Task abstract and Complete A-Enum

Hikoya committed:
Fix Skylark.java with Java coding standard

Hikoya committed:
Move Enum to separate file and create new Printer class

Creating a separate class helps to abstract out code and makes it easier to understand

Hikoya committed:
Create constructor for Skylark

Hikoya committed:
Remove list of commands and place into new Command class

Hikoya committed:
Update Skylark to use this. for object variables

Hikoya committed:
Remove static modifier from response function

Hikoya committed:
Implement reading from file when Skylark starts

Hikoya committed:
Add save functionality to TaskList

Hikoya committed:
Add parsing of date into Deadline and Event

Hikoya committed:
Merge branch 'branch-Level-7'

Hikoya committed:
Merge branch 'branch-Level-8'

Hikoya committed:
Change outputFormat to inputFormat for writing into file

Hikoya committed:
Create new Storage class for file operations

Hikoya committed:
Create new default package Skylark

Hikoya committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

Hikoya committed:
Add Gradle support to Skylark

Hikoya committed:
Add JUnit Tests into Deadline, Event and Todo

Hikoya committed:
Add settings for JAR packaging

Hikoya committed:
Add JavaDocs to all classes, methods

Hikoya committed:
Add code to comply with coding standard

Hikoya committed:
Add Find command into list of commands

Hikoya committed:
Merge branch 'branch-A-JavaDoc'

Hikoya committed:
Merge branch 'branch-A-CodingStandard'

Hikoya committed:
Merge branch 'branch-Level-9'

Hikoya committed:
Add new packages to separate out functions

Hikoya committed:
Add new test for TaskList and Storage

Hikoya committed:
Add JavaDocs into overridden method and edit code to standard

Hikoya committed:
Add support for JavaFX to Gradle

Hikoya committed:
Add FXML file and corresponding Java file for JavaFX

Hikoya committed:
Add Launcher and Main class

Hikoya committed:
Move Skylark to its own package

Hikoya committed:
Update Command.java to return a String for 'run'

Hikoya committed:
Remove Scanner and Printer from Skylark

Hikoya committed:
Add a hello message when user start Skylark

Hikoya committed:
Add close function in JavaFX

Hikoya committed:
Change the user picture to something else

Hikoya committed:
Update Command.java with separator for Find

Hikoya committed:
Merge branch 'branch-Level-10'

normkoh reviewed (commented):
Code is clean and easy to follow. Comments are adequately provided where there may be confusion. There are no coding standard violations (that I could find). Overall, style-wise, I see no problems with the code. Great job!

Hikoya commented:
> Code is clean and easy to follow. Comments are adequately provided where there may be confusion. There are no coding standard violations (that I could find). Overall, style-wise, I see no problems with the code. Great job!

Thank you so much for your review!

Hikoya committed:
Add assertions to program flow

Hikoya committed:
Remove potential magic numbers from Command

Remove potential magic numbers from Command

There are several magic numbers that determine the substring of the command to find out the description of a Task

Separating all these magic numbers improves the code quality.

Let's,
* update the Command class to use constants instead
* modify the code accordingly

Hikoya committed:
Remove emoji from Exception

Command: Remove emoji from Exception description

Emoji is not showing as intended on the GUI

Let's remove all emoji from Exception and return just the text of the issue

paotheroo commented on src/main/java/skylark/command/Command.java:
```diff
@@ -0,0 +1,329 @@
+     * @return A Command object based on interpreted input.
+     */
+    public static Command createCommand(String command) {
+        if (command.equals(CommandList.COMMAND_LIST.toString())) {
+            return new ListCommand(command);
+        } else if (command.length() >= 4 && command.startsWith(CommandList.COMMAND_DONE.toString())) {
+            return new DoneCommand(command);
+        } else if (command.length() >= 6 && command.startsWith(CommandList.COMMAND_UNDONE.toString())) {
+            return new UndoneCommand(command);
+        } else if (command.length() >= 4 && command.startsWith(CommandList.COMMAND_TODO.toString())) {
+            return new TodoCommand(command);
+        } else if (command.length() >= 8 && command.startsWith(CommandList.COMMAND_DEADLINE.toString())) {
+            return new DeadlineCommand(command);
+        } else if (command.length() >= 5 && command.startsWith(CommandList.COMMAND_EVENT.toString())) {
+            return new EventCommand(command);
+        } else if (command.length() >= 6 && command.startsWith(CommandList.COMMAND_DELETE.toString())) {
+            return new DeleteCommand(command);
+        } else if (command.length() >= 5 && command.startsWith(CommandList.COMMAND_FIND.toString())) {
+            return new FindCommand(command);
```
I like the way you checked each case thoroughly using .length() and .startsWith(). A possible alternative could be to read the first element in a String array after using .split(" ")

paotheroo commented on src/main/java/skylark/command/Command.java:
```diff
@@ -0,0 +1,329 @@
+
+        @Override
+        public String run(TaskList taskList) throws SkylarkException {
+            if (super.getInput().equals(CommandList.COMMAND_TODO.toString())) {
+                throw new SkylarkException("The description of a todo cannot be empty.");
+            }
```
Understand that you're checking for empty descriptions here (for deadline and event as well), but what if the input is "todo          " (however many spaces)?

paotheroo reviewed (commented):
Overall, awesome work on your iP Chung Loong 👍

Hikoya committed:
Remove StringBuilder from FindCommand

Hikoya committed:
Merge pull request #1 from Hikoya/branch-A-CodeQuality

Add code quality standards

Hikoya committed:
Merge pull request #2 from Hikoya/branch-A-Assertions

Add assertions to program flow

Hikoya committed:
Add gradle.yml into Github as a workflow

Hikoya committed:
Add new createFile function for abstraction

Hikoya committed:
Add guard clause for get function in TaskList

Hikoya committed:
Add tagging functionality to Tasks

Hikoya committed:
Merge pull request #3 from Hikoya/branch-C-Tagging

Add Tagging functionality to Tasks

Francis-Tan commented on src/main/java/skylark/command/CommandList.java:
```diff
@@ -0,0 +1,37 @@
+
+/** List of commands available for the user. */
+public enum CommandList {
+    COMMAND_BYE, COMMAND_LIST, COMMAND_DONE,
+    COMMAND_UNDONE, COMMAND_DEADLINE,
+    COMMAND_TODO, COMMAND_EVENT, COMMAND_DELETE,
+    COMMAND_FIND, COMMAND_TAG;
```
Maybe try this to make it more compact
```
    COMMAND_BYE("bye"), COMMAND_LIST("list");
    
    private final String name;

    CommandList(String name) {
        this.name = name;
    }

    @Override
    public String toString() {
        return name;
    }
```

Francis-Tan commented on src/main/java/skylark/task/ToDo.java:
```diff
@@ -0,0 +1,48 @@
+package skylark.task;
+
+/** Represents a ToDo Task. */
+public class ToDo extends Task {
```
Nitpick but maybe rename this to Todo to be safe in conforming to coding standard?

Francis-Tan reviewed (commented):
Well done! I find your code design easy to follow and interesting

Hikoya committed:
Update Tasklist.java to follow JavaDoc convention

Hikoya committed:
Change code according to ip feedback

Hikoya committed:
Merge branch 'master' of https://github.com/Hikoya/ip

Hikoya committed:
Change ToDo to Todo

Hikoya committed:
Add title and icon to Skylark window

Hikoya committed:
Add font-bold to user command

Hikoya committed:
Create detailed user guide for Skylark

Hikoya committed:
Update DialogBox to fix Checkstyle errors

Hikoya committed:
Update README.md in main folder to Skylark

Hikoya committed:
Add constant to remove magic numbers in Storage

Hikoya committed:
Update README.md with latest release link

Hikoya committed:
Update README.md to fix bug with markdown
